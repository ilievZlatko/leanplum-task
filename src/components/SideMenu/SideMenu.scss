@import '../../styles/colors';
@import '../../styles/spacings';

.side-menu {
	background-color: #f5f5f5;
	width: 500px;
	height: 100%;
	position: fixed;
	top: 0;
	left: 0;
	bottom: 0;

	&__title {
		display: flex;
		flex-direction: column;
		justify-content: center;
		padding-left: $space-xl;
		height: 120px;
		border-bottom: 1px solid $color-title-border;
	}

	&__items-container {
		padding: $space-xl;
		overflow-y: auto;
		display: flex;
		flex-direction: column;
		height: calc(100% - 120px);
	}
}
