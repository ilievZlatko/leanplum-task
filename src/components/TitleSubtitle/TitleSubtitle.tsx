import React from 'react'

interface Props {
	title: string | React.ReactNode
	subtitle: string | React.ReactNode
}

const TitleSubtitle: React.FC<Props> = ({ title, subtitle, ...otherProps }) => {
	return (
		<div {...otherProps}>
			<h2 data-test='title' className='title'>
				{title}
			</h2>
			<span data-test='subtitle' className='subtitle'>
				{subtitle}
			</span>
		</div>
	)
}

export default TitleSubtitle
