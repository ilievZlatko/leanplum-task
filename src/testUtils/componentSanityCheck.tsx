import React from 'react'
import { shallow } from 'enzyme'

export const componentSanityCheck = (
	label: string,
	Element: React.ReactElement,
) =>
	describe(`Sanity checks for ${label}`, () => {
		it('is defined', () => {
			expect(Element).toBeDefined()
		})

		it('renders the element', () => {
			const wrapper = shallow(Element)
			const expected = false
			const actual = wrapper.isEmptyRender()
			expect(expected).toEqual(actual)
		})

		it('matches the snapshot', () => {
			const wrapper = shallow(Element)
			const expected = wrapper.html()
			expect(expected).toMatchSnapshot()
		})
	})
