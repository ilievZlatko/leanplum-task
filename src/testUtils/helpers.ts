import { ShallowWrapper, ReactWrapper } from 'enzyme'

export const findByTestAttr = (
	component: ShallowWrapper | ReactWrapper | any,
	attr: string,
): ShallowWrapper | ReactWrapper => {
	return component.find(`[data-test='${attr}']`)
}
