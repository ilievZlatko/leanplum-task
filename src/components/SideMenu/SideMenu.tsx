import React, { Children } from 'react'

import SideMenuItem from './SideMenuItem/SideMenuItem'
import { TitleSubtitle } from '../../components'
import './SideMenu.scss'

interface Props {
	children: React.ReactNode
}

const SideMenu = ({ children, ...otherProps }: Props) => {
	const count = Children.count(children)

	return (
		<div className='side-menu' {...otherProps}>
			<div className='side-menu__title'>
				<TitleSubtitle
					title='Users in Audience'
					subtitle={`Total Users - Showing ${count} matching users`}
				/>
			</div>

			<div className='side-menu__items-container'>
				{React.Children.map(children, child => {
					if (!child)
						throw new Error(
							'SideMenu Component needs child elements of type SideMenuItem in order to be used!',
						)

					if (
						React.isValidElement(child) &&
						(child as React.ReactElement).type === SideMenuItem
					) {
						return child
					} else {
						throw new Error(
							'SideMenu Component can only accept child elements of type SideMenuItem',
						)
					}
				})}
			</div>
		</div>
	)
}

SideMenu.SideMenuItem = SideMenuItem

export default SideMenu
