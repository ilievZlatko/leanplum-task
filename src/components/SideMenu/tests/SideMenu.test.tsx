import React from 'react'
import { shallow } from 'enzyme'

import SideMenu from '../SideMenu'
import { componentSanityCheck } from '../../../testUtils/componentSanityCheck'
import { users } from './mocks'
import TitleSubtitle from '../../TitleSubtitle/TitleSubtitle'

describe('SideMenu', () => {
	componentSanityCheck(
		'SideMenu',
		<SideMenu>
			<SideMenu.SideMenuItem
				user={users[0]}
				onClick={jest.fn()}
				active={true}
			/>
			<SideMenu.SideMenuItem
				user={users[1]}
				onClick={jest.fn()}
				active={false}
			/>
			<SideMenu.SideMenuItem
				user={users[2]}
				onClick={jest.fn()}
				active={false}
			/>
		</SideMenu>,
	)

	it('Should display the correct title and subtitle', (done: jest.DoneCallback) => {
		const wrapper = shallow(
			<SideMenu>
				<SideMenu.SideMenuItem
					user={users[0]}
					onClick={jest.fn()}
					active={true}
				/>
				<SideMenu.SideMenuItem
					user={users[1]}
					onClick={jest.fn()}
					active={false}
				/>
				<SideMenu.SideMenuItem
					user={users[2]}
					onClick={jest.fn()}
					active={false}
				/>
			</SideMenu>,
		)

		const titleSection = wrapper.find(TitleSubtitle)

		expect(titleSection.prop('subtitle')).toEqual(
			'Total Users - Showing 3 matching users',
		)
		expect(wrapper.find(SideMenu.SideMenuItem)).toHaveLength(3)
		done()
	})
})
