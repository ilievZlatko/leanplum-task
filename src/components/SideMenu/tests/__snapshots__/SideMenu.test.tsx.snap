// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu Sanity checks for SideMenu matches the snapshot 1`] = `"<div class=\\"side-menu\\"><div class=\\"side-menu__title\\"><div><h2 data-test=\\"title\\" class=\\"title\\">Users in Audience</h2><span data-test=\\"subtitle\\" class=\\"subtitle\\">Total Users - Showing 3 matching users</span></div></div><div class=\\"side-menu__items-container\\"><aside class=\\"menu-item menu-item__active\\" data-test=\\"menu-item\\"><div><h2 data-test=\\"title\\" class=\\"title\\">usr_0</h2><span data-test=\\"subtitle\\" class=\\"subtitle\\"><span>1 Devices - </span><span>2 Sessions - </span><span>Sofia</span></span></div></aside><aside class=\\"menu-item\\" data-test=\\"menu-item\\"><div><h2 data-test=\\"title\\" class=\\"title\\">usr_1</h2><span data-test=\\"subtitle\\" class=\\"subtitle\\"><span>2 Devices - </span><span>5 Sessions - </span><span>San Francisco</span></span></div></aside><aside class=\\"menu-item\\" data-test=\\"menu-item\\"><div><h2 data-test=\\"title\\" class=\\"title\\">usr_2</h2><span data-test=\\"subtitle\\" class=\\"subtitle\\"><span>3 Devices - </span><span>8 Sessions - </span><span>Amsterdam</span></span></div></aside></div></div>"`;
